conan/tools/apple/xcodedeps.py

```diff
@@ -213,28 +217,60 @@ def _content(self):
 
         # Generate the config files for each component with name conan_pkgname_compname.xcconfig
         # If a package has no components the name is conan_pkgname_pkgname.xcconfig
-        # Then all components are included in the conan_pkgname.xcconfig file
+        # All components are included in the conan_pkgname.xcconfig file
         host_req = self._conanfile.dependencies.host
         test_req = self._conanfile.dependencies.test
-        for dep in list(host_req.values()) + list(test_req.values()):
+        all_deps = list(host_req.values()) + list(test_req.values())
+        for dep in all_deps:
+
             dep_name = _format_name(dep.ref.name)
 
             include_components_names = []
             if dep.cpp_info.has_components:
-                for comp_name, comp_cpp_info in dep.cpp_info.get_sorted_components().items():
-                    component_deps = []
-                    for req in comp_cpp_info.requires:
-                        req_pkg, req_cmp = req.split("::") if "::" in req else (dep_name, req)
-                        component_deps.append((req_pkg, req_cmp))
 
-                    component_content = self.get_content_for_component(dep_name, comp_name, comp_cpp_info, component_deps)
+                sorted_components = dep.cpp_info.get_sorted_components().items()
+                for comp_name, comp_cpp_info in sorted_components:
+
+                    # returns: ("list of cpp infos from required components in same package",
+                    #           "list of names from required components from other packages")
+                    def _get_component_requires(component):
+                        requires_external = [(req.split("::")[0], req.split("::")[1]) for req in
+                                             component.requires if "::" in req]
+                        requires_internal = [dep.cpp_info.components.get(req) for req in
+                                             component.requires if "::" not in req]
+                        return requires_internal, requires_external
+
+                    # these are the transitive dependencies between components of the same package
+                    transitive_internal = []
+                    # these are the transitive dependencies to components from other packages
+                    transitive_external = []
+
+                    # return the internal cpp_infos and external components names
+                    def _transitive_components(component):
+                        requires_internal, requires_external = _get_component_requires(component)
+                        transitive_internal.append(component)
+                        transitive_internal.extend(requires_internal)
+                        transitive_external.extend(requires_external)
+                        for require in requires_internal:
+                            _transitive_components(require)
```

Comment:
This recursion is a bit complicated to follow. Maybe the returning of the already outer scope defined lists ``transitive_internal``, ``transitive_external`` is unnecessary?